cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -596,8 +596,15 @@ public boolean accept(String className, URL classUrl, URL classPathUrl) {
    *
    * @return The generated {@link File} in the given target directory
    */
-  private File generateJobJar(File targetDir) throws IOException {
+  private File generateJobJar(File tempDir, boolean isLocal) throws IOException {
+    // in local mode, Spark Streaming with checkpointing will expect the same job jar to exist
+    // in all runs of the program. This means it can't be created in the temporary directory for the run,
+    // but must persist between runs
+    File targetDir = isLocal ? tempDir.getParentFile() : tempDir;
```

Comment:
This is a bit hack to just get the parent. If really for persisting it in local mode, better pass in the `cConf` and persist it to a directory under the data directory (e.g. `data/runtime/spark/emptyJob.jar`).